### PR TITLE
DrivAerML: MLP ratio sweep (ratio=2/3/6)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The MLP ratio (controlling the feedforward expansion factor inside each transformer block) has never been ablated on DrivAerML. The default value of 4 means the MLP hidden dimension is 4×512=2048. With only 394 training batches, this MLP capacity may be creating unnecessary overfitting pressure. Reducing to mlp-ratio=2 (MLP dim=1024) or mlp-ratio=3 (MLP dim=1536) cuts parameters by 25-50% while keeping attention capacity unchanged. Conversely, mlp-ratio=6 (MLP dim=3072) tests whether more MLP capacity helps DrivAerML's complex 3D surface geometry. This is a pure capacity knob that hasn't been pulled and costs nothing in attention compute.

## Baseline
val_primary/surface_rel_l2_pct = **4.619%** (PR #2691, 4L/512d/8H, T_max=30, lr=5e-4, AdamW, WD=1e-2, mlp-ratio=4 default)

## Runs

All runs keep golden config values: 4L/512d/8H, lr=5e-4, WD=1e-2, gc=1.0, T_max=30, AdamW, no-use-ema, enable-fourier, 50k surface points. Only the MLP ratio changes.

**Run 1 (CUDA 0): mlp-ratio=2**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-mlp-ratio 2 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --wandb-group drivaerml-mlp --wandb-name drivaerml-mlp2
```

**Run 2 (CUDA 1): mlp-ratio=3**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-mlp-ratio 3 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --wandb-group drivaerml-mlp --wandb-name drivaerml-mlp3
```

**Run 3 (CUDA 2): mlp-ratio=6**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=2 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-mlp-ratio 6 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --wandb-group drivaerml-mlp --wandb-name drivaerml-mlp6
```

## Expected Outcome
val_primary/surface_rel_l2_pct < **4.619%**

Report val_primary/surface_rel_l2_pct for each run. If --model-mlp-ratio is not a recognized flag, report the error and try --mlp-ratio instead.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT shouko:

## Results

### Summary

| MLP Ratio | MLP Dim | Best Val (surface_rel_l2_pct) | Best Epoch | Final Test | Epochs Run | W&B Run ID | Status |
|-----------|---------|-------------------------------|------------|------------|------------|------------|--------|
| 2 | 1024 | 12.579% | 40 | 68.921% | 244 | k90btjfo | **DIVERGED** (grad norm → Infinity @ ep58) |
| **3** | 1536 | **5.174%** | 243 | 6.544% | 246 | dvd4hkco | Stable |
| **6** | 3072 | **5.069%** | 234 | 7.370% | 242 | usl39efl | Stable |
| 4 (baseline) | 2048 | **4.619%** | 267 | — | 267 | — | Baseline (PR #2691) |

**External target:** <3.71%

### Baseline Comparison

- **mlp-ratio=2:** ❌ Diverged catastrophically. Grad norm exploded to Infinity around epoch 58 and never recovered. Halving MLP capacity destabilizes training completely.
- **mlp-ratio=3:** ❌ Best val 5.174% vs baseline 4.619% — **12.0% worse**. Stable training, but insufficient MLP capacity.
- **mlp-ratio=6:** ❌ Best val 5.069% vs baseline 4.619% — **9.7% worse**. Stable training, slightly better than ratio=3 despite 50% more MLP params.

**None of the tested ratios beat the baseline.** The default mlp-ratio=4 remains optimal.

### Commands Used

```bash
# mlp-ratio=2 (GPU 0)
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-mlp-ratio 2 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --wandb-group drivaerml-mlp --wandb-name drivaerml-mlp2 --agent shouko

# mlp-ratio=3 (GPU 1)
cd target/icml2026 && CUDA_VISIBLE_DEVICES=1 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-mlp-ratio 3 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --wandb-group drivaerml-mlp --wandb-name drivaerml-mlp3 --agent shouko

# mlp-ratio=6 (GPU 2)
cd target/icml2026 && CUDA_VISIBLE_DEVICES=2 SENPAI_MAX_EPOCHS=9999 SENPAI_TIMEOUT_MINUTES=180 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --model-mlp-ratio 6 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --wandb-group drivaerml-mlp --wandb-name drivaerml-mlp6 --agent shouko
```

### What Happened

The MLP ratio has a clear impact on DrivAerML training dynamics:

1. **mlp-ratio=2 is catastrophically unstable.** With only 1024-dim MLP, training diverges around epoch 58. The reduced MLP capacity creates a bottleneck that causes gradient explosion. This is a hard lower bound — ratio=2 is too small for 512d attention.

2. **mlp-ratio=3 and 6 both train stably but converge slower.** Despite reaching ~240 epochs (close to baseline's 267), neither achieves the baseline's 4.619%. The convergence trajectories suggest they would still lag even with extended training.

3. **mlp-ratio=4 is a sweet spot.** The default 2048-dim MLP provides the right balance of capacity and regularization for DrivAerML's 394-batch training regime. Both smaller (ratio=3) and larger (ratio=6) ratios underperform.

4. **Interestingly, ratio=6 slightly outperforms ratio=3** (5.07% vs 5.17%), suggesting that within the stable regime, more MLP capacity helps marginally — but the jump from 4→6 doesn't compensate for whatever regularization benefit ratio=4 provides.

### Peak Memory

All runs fit comfortably on single H100 GPUs (batch_size=1, 50k surface points).

### Suggested Follow-ups

- **mlp-ratio=5 is the only untested adjacent value** — unlikely to beat ratio=4 given the trend, but could confirm the optimality boundary.
- The divergence at ratio=2 suggests a **minimum MLP capacity threshold** for DrivAerML. This may interact with model depth — 3L/512d might tolerate lower ratios.
- The result confirms the current golden config's ratio=4 is well-chosen. Future DrivAerML experiments should keep this fixed.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2857 because ADVISOR: CLOSE — MLP ratio sweep conclusive. ratio=2 diverged, ratio=3 (5.17%), ratio=6 (5.07%) both worse than baseline 4.619%. ratio=4 confirmed optimal..
